src/components/calcite-input/calcite-input.tsx

```diff
@@ -0,0 +1,366 @@
+
+  @Element() el: HTMLElement;
+
+  //--------------------------------------------------------------------------
+  //
+  //  Properties
+  //
+  //--------------------------------------------------------------------------
+
+  /** specify the status of the input field, determines message and icons */
+  @Prop({ mutable: true, reflect: true }) status: "invalid" | "valid" | "idle";
+
+  /** specify if the input is in loading state */
+  @Prop({ mutable: true, reflect: true }) loading: boolean = false;
+
+  /** specify the scale of the input, defaults to m */
+  @Prop({ mutable: true, reflect: true }) scale: "s" | "m" | "l" = "m";
+
+  /** specify the appearance type - minimal or default */
+  @Prop({ mutable: true, reflect: true }) appearance: "minimal" | "default" =
+    "default";
+
+  /** specify the width */
+  @Prop({ mutable: true, reflect: true }) width: "auto" | "half" | "full" =
+    "full";
+
+  /** optionally add prefix  **/
+  @Prop({ mutable: true }) prefixText?: string;
+
+  /** optionally add suffix  **/
+  @Prop({ mutable: true }) suffixText?: string;
+
+  /** optionally pass icon path data - pass only raw path data from calcite ui helper  */
+  @Prop({ mutable: true, reflect: true }) icon?: string;
+
+  /** specify the placement of the number buttons */
+  @Prop({ mutable: true, reflect: true }) numberButtonType?:
+    | "vertical"
+    | "horizontal"
+    | "none" = "vertical";
+
+  /** specify the alignment of dropdown, defaults to left */
+  @Prop({ mutable: true, reflect: true }) theme: "light" | "dark" = "light";
+
+  /** is the input required */
+  @Prop() required: boolean = false;
+
+  /** should the input autofocus */
+  @Prop() autofocus: boolean = false;
+
+  //--------------------------------------------------------------------------
+  //
+  //  Lifecycle
+  //
+  //--------------------------------------------------------------------------
+
+  connectedCallback() {
+    // validate props
+    let statusOptions = ["invalid", "valid", "idle"];
+    if (!statusOptions.includes(this.status))
+      this.status = getElementProp(this.el, "status", "idle");
+
+    let theme = ["light", "dark"];
+    if (!theme.includes(this.theme)) this.theme = "light";
+
+    let scale = ["s", "m", "l"];
+    if (!scale.includes(this.scale)) this.scale = "m";
+
+    let width = ["auto", "half", "full"];
+    if (!width.includes(this.width)) this.width = "full";
+
+    let appearance = ["minimal", "default"];
+    if (!appearance.includes(this.appearance)) this.appearance = "default";
+
+    let numberButtonType = ["vertical", "horizontal", "none"];
+    if (
+      this.inputType === "number" &&
+      !numberButtonType.includes(this.numberButtonType)
+    )
+      this.numberButtonType = "vertical";
+  }
+
+  componentDidLoad() {
+    this.childEl = this.el.querySelector(`${this.childElType}`);
+  }
+
+  componentWillLoad() {
+    this.inputType = this.el.getAttribute("type");
```

Comment:
Could this be done via properties and not attributes? Reasoning is that Stencil components focus more on props than attributes. Also, by not having props, the compiler won't be able to generate proper types.

Reply:
Yep good catch.

Reply:
Adding all "reasonable" options as allowable types (ie we probably don't need to allow checkbox, or radio, or range, since other components handle those (or, will in the case of radio)).